Read serial line in chunks instead of character by character

wentasah opened the pull request:
This significantly reduces CPU load caused by this driver.

Without this change, CPU load on a powerful x86 machine with a USB-connected GPS is 36%. The output of the strace command looks like this:

    ...
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997160})
    read(33, "$", 1)                        = 1
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997160})
    read(33, "G", 1)                        = 1
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997160})
    read(33, "P", 1)                        = 1
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997180})
    read(33, "R", 1)                        = 1
    ...

It can be seen that characters are read from the serial line one by one. This is due to how IOBase.readline() interacts with pySerial.

In this commit, we implement the same functionality as IOBase.readline(), but in a way allowing to read the serial line characters in bigger chunks. With it, CPU load decreases to 7.5% and strace reports this:

    ...
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=926286000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=906510472})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 422) = 143
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=906287000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=886945652})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 279) = 143
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=886721000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=866882570})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 136) = 136
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999996000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999991549})
    read(33, ",,*57\r\n", 1024)             = 7
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999827000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=985949688})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 1017) = 143
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=985746000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=966235438})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 874) = 143
    ...

Reading from the kernel happens most of the time in chunks of 143 bytes (for our GPS) and therefore, the system call overhead is 143× lower.

Tested on Septentrio mosaic-H GPS.

wentasah committed:
Read serial line in chunks instead of character by character

This significantly reduces CPU load caused by this driver.

Without this change, CPU load on a powerful x86 machine with a
USB-connected GPS is 36%. The output of the strace command looks
like this:

    ...
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997160})
    read(33, "$", 1)                        = 1
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997160})
    read(33, "G", 1)                        = 1
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997160})
    read(33, "P", 1)                        = 1
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999999000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999997180})
    read(33, "R", 1)                        = 1
    ...

It can be seen that characters are read from the serial line one by
one. This is due to how IOBase.readline() interacts with pySerial.

In this commit, we implement the same functionality as
IOBase.readline(), but in a way allowing to read the serial line
characters in bigger chunks. With it, CPU load decreases to 7.5% and
strace reports this:

    ...
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=926286000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=906510472})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 422) = 143
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=906287000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=886945652})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 279) = 143
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=886721000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=866882570})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 136) = 136
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999996000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=999991549})
    read(33, ",,*57\r\n", 1024)             = 7
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=999827000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=985949688})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 1017) = 143
    pselect6(35, [33 34], [], [], {tv_sec=1, tv_nsec=985746000}, NULL) = 1 (in [33], left {tv_sec=1, tv_nsec=966235438})
    read(33, "$GPRMC,,V,,,,,,,,,,N*53\r\n$GPGGA,"..., 874) = 143
    ...

Reading from the kernel happens most of the time in chunks of 143
bytes (for our GPS) and therefore, the system call overhead is 143×
lower.

Tested on Septentrio mosaic-H GPS.

evenator commented on src/libnmea_navsat_driver/nodes/nmea_serial_driver.py:
```diff
@@ -50,17 +51,22 @@ def main(args=None):
-                        "Please report this issue at github.com/ros-drivers/nmea_navsat_driver, including a bag file "
-                        "with the NMEA sentences that caused it." % e)
+                data.extend(GPS.read(1024)) # read at most 1024 bytes
+                lines = data.splitlines(keepends=True)
+                # process complete lines
+                for line in lines[0:-1]:
+                    line = line.decode("utf-8").rstrip()
+                    try:
+                        driver.add_sentence(line, frame_id)
+                    except ValueError as e:
+                        driver.get_logger().warn(
+                            "Value error, likely due to missing fields in the NMEA message. Error was: %s. "
+                            "Please report this issue at github.com/ros-drivers/nmea_navsat_driver, including a bag file "
+                            "with the NMEA sentences that caused it." % e)
+                # continue with the last, incomplete line
+                data = lines[-1]
```
I think there's an edge case here. If the byte array contains only complete lines, the last element will be held in the data buffer until the next GPS read completes, which introduces unnecessary latency in parsing that line.

I think this would work?
```suggestion
                lines = data.splitlines(keepends=True)
                data.clear()
                def process_line(line):
                    line = line.decode("utf-8").rstrip()
                    try:
                        driver.add_sentence(line, frame_id)
                    except ValueError as e:
                        driver.get_logger().warn(
                            "Value error, likely due to missing fields in the NMEA message. Error was: %s. "
                            "Please report this issue at github.com/ros-drivers/nmea_navsat_driver, including a bag file "
                            "with the NMEA sentences that caused it." % e)
                # process complete lines
                for line in lines[0:-1]:
                    process_line(line)
                if lines[-1].endswith('\r\n'):
                    process_line(lines[-1])
                else:
                    # continue with the last, incomplete line
                    data = lines[-1]
```

wentasah replied:
Right. I only changed `lines[-1].endswith('\r\n')` to `lines[-1].endswith(b'\r\n')` in your suggestion and tested that the code works. PR updated. Thanks for noticing.

evenator commented:
I'm testing this locally and seeing some pretty problematic behavior due to the interaction of the timeout (2 seconds) and the buffer length (1024 char) for my device. The call to `read()` will always wait for 1024 characters or the full two seconds, which can hold several complete sentences in the buffer, introducing a lot of latency (i.e. up to 2 seconds).

Would `GPS.read_until(size=1024)` achieve the same performance improvement while avoiding the buffering latency?

wentasah committed:
Add non-broken implementation of pySerial read()

This commit copies the implementation of seria.Serial.read() and
removes one line from it: "if timeout.expired():". With this change,
the function provides classical POSIX semantics of read(), which
returns as soon as some bytes are received. The implementation in
pySerial either waits for the full buffer or timeout expiration -
nothing in between. This means that with pySerial, you either pay big
overherd by reading character-by-character or have increased latency
due to waiting.

Unfortunately, there is no way to fix this in pySerial without
changing the semantics and possibly breaking some applications or
introducing a fixed function with a different name. Also, it seems
that pySerial is no loger maintained (last commit 16 months ago, PRs
without maintainer responses).

It would be much simpler not to use pySerial and use /dev/ttyXXX
directly. This project does not need the features offered by pySerial
(mainly abort read from another thread). The only problem of not using
pySerial would be lower portability.

wentasah commented:
Unfortunately, `GPS.read_until` would have the same overhead. I tried to come up with a better way of using pySerial API and there is no way how to use it without either overhead or latency.

In the new commit, I copied the implementation of `read()` from pySerial and fixed the problem by removing the line. With it, the both overhead and latency is low. I copy my commit message here:

> This commit copies the implementation of seria.Serial.read() and removes one line from it: "if timeout.expired():". With this change, the function provides classical POSIX semantics of read(), which returns as soon as some bytes are received. The implementation in pySerial either waits for the full buffer or timeout expiration - nothing in between. This means that with pySerial, you either pay big overherd by reading character-by-character or have increased latency due to waiting.
>
> Unfortunately, there is no way to fix this in pySerial without changing the semantics and possibly breaking some applications or introducing a fixed function with a different name. Also, it seems that pySerial is no loger maintained (last commit 16 months ago, PRs without maintainer responses).
>
> It would be much simpler not to use pySerial and use /dev/ttyXXX directly. This project does not need the features offered by pySerial (mainly abort read from another thread). The only problem of not using pySerial would be lower portability.


What do you think? Do you want to use pySerial to stay compatible with many OSes? If you care only about Linux, I'd suggest getting rid of pySerial. I'll provide the implementation.

ggaessler commented:
Wouldn't a good and simple solution be to just set the timeout to zero and the read size to something like the proposed 1024? That should be fine for performance and latency. That's also the way I used it for quite some years by now without any problems:
https://github.com/boschresearch/nmea_navsat_driver/blob/master/src/libnmea_navsat_driver/nodes/nmea_serial_driver.py#L177-L186
EDIT: The line separation would still be needed. I don't need it because I anyway process mixed binary and ASCII data from the receiver since I also enabled proprietary messages.

wentasah commented:
I think that this way, you would end up busy waiting for the serial line. The CPU load would always be 100%. At least this is what I saw when digging through the source code. Is your experience different?

evenator commented:
What about a timeout that's fast in GPS terms but slow in CPU terms? Would
10ms be an acceptable compromise?

On Wed, Apr 19, 2023, 3:37 PM Michal Sojka ***@***.***> wrote:

> I think that this way, you would end up busy waiting for the serial line.
> The CPU load would always be 100%. At least this is what I saw when digging
> through the source code. Is your experience different?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ros-drivers/nmea_navsat_driver/pull/164#issuecomment-1515275155>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAEKUHEPB6EASGJ4SQRALVDXCA5IPANCNFSM6AAAAAAWP3V5WI>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***>
>

ggaessler commented:
Sorry my changes were a long time back and I didn't check the full code again. Yes, there indeed was some additional sleep for a time short in GNSS, but long in CPU terms, by default 10 ms, just as @evenator proposed.
I've never tried setting the timeout parameter that way, but that could be the cleaner solution compared to my timeout zero plus sleep.

evenator commented:
@wentasah Are you still working on this enhancement, or should I close the PR as abandoned?

wentasah commented:
I'll continue working on this in upcomming weeks.
